### DISCUSSION TITLE
1 second delay between every data packet in websocket live streaming

### DISCUSSION THREAD

**Discussion by anunayajoshi** [deepgram/community#426] [General help] [answered]
Hi, so I'm trying to streamline the STT -> LLM -> TTS pipeline, and we've managed to get the LLM and synthesis all down to sub-second latencies, which are great, but the interrupt system hasn't been working well.

And I realised, that between every data packet in the websocket from deepgram, it takes around a second for the transcription to come in. Here's a screenshot, where I simply timed the difference in time between each data packet received. 
Here's when silent:
![image](https://github.com/deepgram/community/assets/47421902/9906a48a-78de-4d72-a524-1f3e78be885c)

It does speed up to 0.1/0.2 seconds sometimes when talking, but for the most part, it's still around 1 second as seen here, which causes my other pipeline to interject and synthesize while the user is talking:
![image](https://github.com/deepgram/community/assets/47421902/474cd7e1-e166-40a8-91d0-6062612ed41d)

As such, can I understand what I can do to optimize the latency for the data packets coming in? Is that the expected latency for everyone, or am I mistaken in my implementation?

**Comment by nikolawhallon** [comment]
This doc might be the most helpful: https://developers.deepgram.com/docs/understand-endpointing-interim-results

If interim results are turned on, Deepgram will send an interim result about once per second (you are seeing this with your logs there, most of those are interim results as `is_final` is `False`). Deepgram usually sends final results every 3-5 seconds, unless our endpointing algorithm kicks in - the doc explains the algorithm, but in short, it helps return finalized transcripts back to you after someone has finished speaking - it does this by inspecting the audio and noticing when there are X ms of silence following speech (by default X is 10, but if you use `endpointing=100` as a query parameter, it will wait for 100 ms of silence following speech before triggering). Messages that have been chunked/generated in this way will have `speech_final: True` in the returned json. Latency for these types of messages is usually X ms (10 by default) + O(100)ms.

Let me know if this helps.

**Reply by anunayajoshi** [reply]
Thanks for the quick response Nikola! Yeah, I'm familiar with the interim results. I guess my question is more of whether is there a way to speed up the rate of the interim results coming in? 

Since given my case of the STT -> LLM -> TTS pipeline being able to synthesize text in less than 1 second, and in the scenario that the STT gives a speech_final too early, and the human is still talking, I want to be able to cancel the synthesis when the new interim results comes in with more transcription. However, what I am finding currently is that I am cancelling my synthesis too late, because the interim results takes 1 second to come in after the previous one, (which means I start synthesizing before the next interim result comes in with a transcription, after which I am able to cancel. 

I either have 2 options, which is to increase endpointing which is a quick fix, (but would negatively affect the realness, since it adds latency to my pipeline), 
or somehow increase the frequency of the interim results, so even if the speech_final is set to True too early, my system can still cancel the synthesis with a new interim result coming in with new transcription. 

Do you think hosting the deepgram model locally would help reduce the latency there, or do you have any other thoughts about this? Happy to discuss! If I'm not wrong, this seems like the bottleneck in my system with the biggest latency, but definitely isn't a simple change to reduce the latency. Would be great to discuss ways to reduce it.

**Reply by nikolawhallon** [reply]
In theory, the correct approach is to set a higher endpointing value, say, 500ms or so - the would simulate the AI acting as a human would - wait for a pause in the other person's speech before responding - so if LLMs and TTS took 0 time to form a response, this would be the best way (imo).

The issue I guess is that LLM+TTS does take time to form a response - therefore, if you were to start computing that response as early as possible, potentially even before you are 100% sure that the speaker has finished speaking (say, less than 500ms), then you could send that response back to the speaker in a reasonable amount of time. The issue is, the speaker can of course resume speaking at any point in time, and you would need now to know /that/ in order to cut off/cancel the LLM+TTS response (I think this is sometimes called barge-in?).

If you have Deepgram on-prem it can be totally reasonable to set very small intervals for interim results, say 200ms, and this could totally help. There is one more potential feature that could do this even better, and get resolutions of ~< 100ms without the expense of interim results - that is, the reverse of endpointing. We currently use an audio-based VAD to determine when someone has started speaking, and then when they have finished speaking - when someone finishes speaking, we send back that `speech_final` message - well what if we, before sending back that message, immediately sent a message when someone /starts/ speaking? It sounds like that would solve your problem?

**Reply by anunayajoshi** [reply]
Thanks for understanding the problem so deeply Nikola, you asked great questions! Yes! I was looking for more of a VAD than the interim results, and have been using the interim results as a workaround to detect when the human has started speaking. 

I'd imagine it would definitely be what I am looking for. I have integrated different open-source VADs like silero VAD and webrtcVAD, but they come with false positives, which clashed with deepgram's own transcription, which was doing fairly well in not detecting those false positives. So I think it would be beneficial to use deepgram's VAD itself, if it's possible to expose that data of when the user started speaking or not!

I'm not too sure if your system also suffers from false positives from the VAD, or it has been handling it rather well. But yes, if there's any way we can get a data response telling us when the human started speaking, we can process our barge-in very easily!

**Reply by juberti** [reply]
Agree exposing the internal VAD state would be useful. It would still be useful to get additional partial transcripts though to better understand whether an "interruption" is really just a continuer, e.g., someone saying "uh huh" or "yeah".

**Reply by nikolawhallon** [reply]
Sorry I didn't reply at the time, but makes sense, I understand the issue and what you are looking for, and I think we can make a feature for that - no promises, but I've been mocking it out in my free time and it seems to be working locally. I may or may not have explained our utterance end feature, which spits out streamed results with the shape `{"type":"UtteranceEnd"}` - I am thinking of having the streaming API spit out similar messages with a shape like `{"type":"SpeechStarted"}` what do you think?

**Reply by anunayajoshi** [reply]
Hey @nikolawhallon , sorry for the delay in my response as well, was out. I think getting that response in the same shape as the utteranceend would work great. Any way I can collaborate with you to test it out and see if it solves the issue?

I have been testing out azure transcriber's VAD, which I think a lot of people don't seem to notice that they expose, but there are some inconsistencies in results I expect from there. Would love to discuss in more detail properly with you.

**Reply by nikolawhallon** [reply]
I have something working, but it would take some time to deploy as a feature to prod. I'm interested in what information Azure exposes about it's VAD - whether it's audio-frame by audio-frame information, or just event information (speaker started, speaker stopped, speaker started, etc). If I were to get this on Deepgram's beta endpoint next week, would you be interested in testing then?

**Reply by anunayajoshi** [reply]
Yup its just event-based information. 

here's the event - https://learn.microsoft.com/en-us/python/api/azure-cognitiveservices-speech/azure.cognitiveservices.speech.recognizer?view=azure-python#azure-cognitiveservices-speech-recognizer-speech-end-detected

they also have their own endpointing functionality, which they don't explicitly say in their python documentation - speechserviceconnection-endsilencetimeoutms

https://learn.microsoft.com/en-us/python/api/azure-cognitiveservices-speech/azure.cognitiveservices.speech.propertyid?view=azure-python#azure-cognitiveservices-speech-propertyid-speechserviceconnection-endsilencetimeoutms

here's the list in more detail

https://learn.microsoft.com/en-us/dotnet/api/microsoft.cognitiveservices.speech.propertyid?view=azure-dotnet


anyways, yes for sure i'd love to test it out on deepgram beta. can i assume, when the speech_final is sent, it also would be roughly when the speech_end event fires right? that's where azure's VAD was behaving unexpectedly.

**Reply by anunayajoshi** [reply]
Hey @nikolawhallon , just checking in. If you want to contact me via email instead - anunayajoshi@gmail.com

**Reply by tmylk** [reply]
I am also interested in this feature.

**Reply by nikolawhallon** [reply]
it will require some PR review and deployment stuff, it's working its way through

**Reply by cateas** [reply]
@nikolawhallon we are interested in this feature as well. We are also subscribed to Premium plan, is it possible to get access to it as well?

Thank you!

**Reply by nikolawhallon** [reply]
The feature is still in PR reviews, it may make its way through by the end of the holidays. I'll update here as it moves along.

**Reply by skyfallsin** [reply]
Interested in this feature as well! Looking forward to release.

**Reply by nikolawhallon** [reply]
I'll update to say that it is in our codebase, but it will need some deploy(s), probably first to beta. Will update when that's done.

**Reply by sb1992** [reply]
interested in this as well. Exactly the same issue. Would love to test it in beta

**Reply by juberti** [reply]
Great to hear this is moving along. @nikolawhallon, can you comment regarding what sort of VAD info will be available in the initial version?

**Reply by nikolawhallon** [reply]
this is not fully deployed yet, but getting there, the API will be `vad_events=true` (like as a query parameter - I think SDK support would follow soon after, the SDK team is planning on it in a future release) and you will get text websocket messages in the form:
```
{"type":"SpeechStarted","channel":[0,1],"timestamp":2.19}
```
so you know what channel this event occurred on (in the same channel format as other Deepgram responses, a tuple of channel number and total number of channels), as well as the timestamp when speech started - we can use this timestamp to measure the latency of this message too (and compare it to, say, when the next ASR result actually comes in - I notice this event fires up to several seconds earlier than the next ASR message, which is the intention)

will stress this is just a feature that the community (here) proposed and I really like, so it'll be beta for sure, and feedback would be awesome for us to formalize it

but again, not fully deployed (it will be deployed along-side other DG updates at the appropriate time)

**Reply by sb1992** [reply]
Hi @nikolawhallon 
Thanks heaps for the update. That sounds perfect. Can you give like an estimate when we could expect these updates coming out?  We have managed to reduce latency from other components significantly and this is the last part left for it and woulld be very helpful to plan our product roadmap for next few weeks.

Cheers

**Reply by skyfallsin** [reply]
@nikolawhallon thank you! please let us know when it's live, we're waiting.

**Reply by nikolawhallon** [reply]
I am on vacation but I think it has been deployed, if you would like to try it out.

**Reply by jjmaldonis** [reply, marked as answer]
It has been deployed, you can find the docs here: https://developers.deepgram.com/docs/start-of-speech-detection#using-speechstarted

**Reply by cateas** [reply]
@jjmaldonis  could you please let us know if you have updated the Deepgram JS SDK with this?

Thank you!

**Reply by jjmaldonis** [reply]
Yes can do @cateas

**Reply by jjmaldonis** [reply]
@cateas it looks like our team was already working on it and the PR was put up an hour ago. You can follow along here to see when it's merged: https://github.com/deepgram/deepgram-js-sdk/pull/235  A new release will likely be made today or within a few days.

**Reply by anunayajoshi** [reply]
Thank you! Do you think it's possible to provide calibration parameter for the VAD? As I am encountering quite a number of false positives, as it may be too sensitive to sound. @jjmaldonis @nikolawhallon

**Reply by nikolawhallon** [reply]
It is not possible to tune currently, but we are getting feedback for the feature now, and this feedback is noted as well.
